docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+- [Typing improvements](https://github.com/algorand-devrel/beaker/pull/147)
+- [Removed inline imports](https://github.com/algorand-devrel/beaker/pull/148)
+- [Removed dead code](https://github.com/algorand-devrel/beaker/pull/149/files)
+- [Added automated release management and versioning](https://github.com/algorand-devrel/beaker/pull/161)
+
+In addition, there are a remaining set of more major (breaking) changes that are recommended. The recommendations are split into 2 categories, recommendations for immediate improvement (i.e. included in v1.0) and future suggestions that can be addressed post v1.0 launch.
+
+The recommended additional areas for immediate improvement are:
+
+- **Replace the class-based structure with an instance based one** - remove some areas of potential surprise for developers and simplify the Beaker codebase by moving to a composable instance-based structure rather than a static class-based structure
+- **Defer PyTEAL compilation** - improve flexibility and future contract output stability by deferring PyTEAL compilation (i.e. Beaker -> TEAL transpilation) to not happen when the Beaker contract is initialised
+- **Renamings** - There are some clear parameters that make sense to rename for various reasons
+- **Key decorator improvements** - Refactor some of the Beaker decorators to fix some bugs and improve user experience
+- **Beaker state refactor** - Refactor of the Beaker state interfaces to improve user extensibility and significantly simplify the Beaker codebase to improve maintainability
+
+The recommended areas for future improvement are:
+
+- Typed client generation from app spec to improve deploy-time and run-time dev experience
+- `Tmpl` values in app spec so you can have type-safe deployment clients that substitute any template values reliably at contract deploy time
+- Refactor storage types (blob, reserved, etc.) to allow use of in-built Python types and operators (terser, more intuitive)
+- Box storage implementation improved to match local/global behaviour and also automatically delete itself on contract deletion
+- Composable and stackable authorization and `@authorize` as a standalone decorator
+- PyTEAL typings to be improved to support types beyond `Expr` where a more explicit type can be specified (improves typing and extensibility)
+- Support referencing an app/lsig via ID/address (deployed separately, potentially automatically as part reading a Directed Acyclic Graph (DAG) in application.json of application dependencies) or bytes (deployed inline, what was previously called precompile, noting this would be deploy-time substitution, not smart contract run-time substitute like `TemplateVariable`), this also may allow precompile to be deprecated (it's a very complex implementation for what we believe to be an advanced edge case)
```

Comment:
I wonder if you've seen this open PR and if it might impact your thinking https://github.com/algorand/go-algorand/pull/4737

Reply:
I had a brief look, and maybe I'm missing some subtler details, but would using macros not just be another way achieve the same outcome?

Reply:
This is a future item anyway, but good to know re: macros. Reminds me of C/C++ macros!